rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMCore.java

```diff
@@ -441,33 +468,47 @@ private RMNode getNodebyUrl(String url) {
     }
 
     /**
-     * Set a node's state to free, after a completed task by it. Set the to free
-     * and move the node to the internal free nodes list. An event informing the
-     * node state's change is thrown to RMMonitoring.
+     * Change the state of the node to free, after a Task has been completed by the specified node.
+     * The node passed as parameter is moved to the list of nodes which are eligible for scheduling if it is not locked.
+     * In all cases, an event informing the node state's change is propagated to RMMonitoring.
      * 
      * @param rmNode node to set free.
      * @return true if the node successfully set as free, false if it was down before.
      */
     private BooleanWrapper internalSetFree(final RMNode rmNode) {
+        if (logger.isDebugEnabled()) {
+            logger.debug("Current node state " + rmNode.getState() + " " + rmNode.getNodeURL());
+            logger.debug("Setting node state to free " + rmNode.getNodeURL());
+        }
+
         // If the node is already free no need to go further
-        logger.debug("Current node state " + rmNode.getState() + " " + rmNode.getNodeURL());
-        logger.debug("Setting node state to free " + rmNode.getNodeURL());
         if (rmNode.isFree()) {
             return new BooleanWrapper(true);
         }
+
         // Get the previous state of the node needed for the event
         final NodeState previousNodeState = rmNode.getState();
+
+
         Client client = rmNode.getOwner();
         if (client == null) {
             // node has been just configured, so the user initiated this action is the node provider
             client = rmNode.getProvider();
         }
-        // reseting owner here
+
+        // resetting owner here
         rmNode.setFree();
```

Comment:
The main issue is with the naming of the `freeNodes` instance variable. As commented on its declaration, this variable is a list that contains nodes which are _free and not locked_, not nodes which are _free_ only. 

I have renamed _freeNodes_ to _eligibleNodes_ along with a comment that repeats the conditions to satisfy when an `add` or `remove` is performed.

I prefer not to introduce a new collection since it will increase the complexity in order to find nodes that satisfy the eligibility condition.

At the end, the best would be to rename `*FreeNodes*` to `*EligibleNodes*` everywhere but it will break the API since the change will be reflected on the REST service output due to Dozer bean mapping.